### PR TITLE
Feat/transactions

### PR DESCRIPTION
**Pull request by alinauroz**
This PR fixes https://github.com/cobudget/cobudget/issues/439

### CODE REVIEW

**Comment by aerugo** [comment]
@alinauroz 
Transaction table currently needs an app reload to update.
Could you update the PR so that changes are also stored in the client?
Check how this has been done in other places in the app and follow that pattern.

I am also asking @gustavlrsn to review this PR.

**Comment by aerugo** [comment]
@alinauroz wanted some clarifications on the issues, so I documented it here: https://www.loom.com/share/14f72b90c76e473faa1b54ca1ee56496

File: ui/styles.css [hunk 1]
```diff
@@ -103,3 +103,7 @@ h5 {
 .markdown a.mention {
   @apply pointer-events-none cursor-default;
 }
+
+.text-xxs {
```
Comment (gustavlrsn) [inline ui/styles.css:107]: This should be done by "extending" the font-size in tailwind.config.js

Comment (alinauroz) [reply ui/styles.css]: done

**Comment by gustavlrsn** [comment]
Looks good overall! 

Having to do the raw SQL query is a bit unfortunate since it won't give us the nice typed features of Prisma, but since we are doing this union it makes sense. Maybe we can figure out a more robust Transaction table that does both things later :).

**Comment by alinauroz** [comment]
> Looks good overall!
> 
> Having to do the raw SQL query is a bit unfortunate since it won't give us the nice typed features of Prisma, but since we are doing this union it makes sense. Maybe we can figure out a more robust Transaction table that does both things later :).



Yes I agree. We had to do somethings manually. But I think overall this design  is much simpler.

File: ui/server/prisma/migrations/20220127203604_new_allocation_columns/migration.sql [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  - Added the required column `amountBefore` to the `Allocation` table without a default value. This is not possible if the table is not empty.
+
+*/
+-- CreateEnum
+CREATE TYPE "AllocationType" AS ENUM ('ADD', 'SET');
+
+-- AlterTable
+ALTER TABLE "Allocation" ADD COLUMN     "allocatedById" TEXT NOT NULL,
+ADD COLUMN     "allocationType" "AllocationType" NOT NULL,
+ADD COLUMN     "amountBefore" INTEGER NOT NULL;
```
[the remark this answers is not in the document]
Comment (aerugo) [reply ui/server/prisma/migrations/20220127203604_new_allocation_columns/migration.sql:15]: I would say: Make them optional first, do the migration, then set all past records amountBefore = 0 and AllocationType = "ADD". Then do another migration to make them required. Finally, find the handfull of records in production where it actually matters to get it right and set those manually.

File: ui/server/prisma/schema.prisma [hunk 3]
```diff
@@ -54,11 +54,12 @@ model CollectionMember {
   createdAt  DateTime @default(now())
   updatedAt  DateTime @updatedAt
 
-  buckets       Bucket[]
-  allocations   Allocation[]
-  contributions Contribution[]
-  comments      Comment[]
-  flags         Flag[]
+  buckets           Bucket[]
+  allocatedBy       Allocation[] @relation("allocatedBy")
```
[the remark this answers is not in the document]
Comment (gustavlrsn) [reply ui/server/prisma/schema.prisma:58]: As opposed to allocations made to that member

[the remark this answers is not in the document]
Comment (aerugo) [reply ui/server/prisma/schema.prisma:58]: So all good to resolve this @gustavlrsn ?

File: ui/server/graphql/schema/index.js [hunk 4]
```diff
@@ -482,9 +488,31 @@ const schema = gql`
     collection: Collection!
     collectionMember: CollectionMember!
     amount: Int!
+    allocatedById: Int!
+    allocationType: AllocationType!
+    amountBefore: Int!
     createdAt: Date
   }
 
+  type CollectionTransaction{
```
[the remark this answers is not in the document]
Comment (gustavlrsn) [reply ui/server/graphql/schema/index.js:497]: Ideally we should then adapt it. The interface `Transaction` are for those values that are shared by both `Allocation` and `Contribution`. But I'm fine with merging this as is. I think we need to adapt the whole transaction thing soon anyway to be able to handle transactions and balances on orgs and collections.

File: ui/server/prisma/migrations/20220127203604_new_allocation_columns/migration.sql (hunk 2 above)
[the remark this answers is not in the document]
Comment (aerugo) [reply ui/server/prisma/migrations/20220127203604_new_allocation_columns/migration.sql:15]: @alinauroz - you can implement the solution above

File: ui/server/graphql/schema/index.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (alinauroz) [reply ui/server/graphql/schema/index.js:497]: @gustavlrsn allocations and contributions have different columns. 

For example, allocation have a column named *allocatedBy* but contributions does not have this column. Similarly there are columns in contributions which are not in allocations.

File: ui/server/prisma/migrations/20220127203604_new_allocation_columns/migration.sql (hunk 2 above)
[the remark this answers is not in the document]
Comment (alinauroz) [reply ui/server/prisma/migrations/20220127203604_new_allocation_columns/migration.sql:15]: > I would say: Make them optional first, do the migration, then set all past records amountBefore = 0 and AllocationType = "ADD". Then do another migration to make them required. Finally, find the handfull of records in production where it actually matters to get it right and set those manually.

Isn't this the same setting default value = 0? Problem will be with allocatedBy. There is a constraint on allocatedById.

**Comment by gustavlrsn** [comment]
Merged in #553

**Pull request closed without merging**